### PR TITLE
Easy Deploy on Save

### PR DESCRIPTION
**Pull request by allileong**
### What does this PR do?
When the workspace setting for deploy-on-save is enabled, changes made to files inside of a package directory will trigger an automatic push or deploy, depending on the type of org that is connected to the workspace. 

**Note:** This functionality is not working in Windows due to a problem with the GlobPattern and I have followed the following issue with VS Code: https://github.com/Microsoft/vscode/issues/65240

### What issues does this PR fix or reference?
W-5552627, #577

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+export enum FileEventType {
+  Create,
+  Change,
+  Delete
+}
+
+const WAIT_TIME_IN_MS = 50;
+
+export async function registerPushOrDeployOnSave() {
+  if (sfdxCoreSettings.getPushOrDeployOnSaveEnabled()) {
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts]: We only register the file watcher if the push-or-deploy-on-save setting is already set to true on core activation. This means that if a user wants to turn on the feature, they will have to restart VS Code. 

Disablement of the feature will be respected immediately will not require a restart of the extension.

File: packages/system-tests/assets/sfdx-simple/.vscode/settings.json [hunk 2]
```diff
@@ -1,3 +0,0 @@
-{
```
Comment (lcampos) [inline packages/system-tests/assets/sfdx-simple/.vscode/settings.json]: Why do we need to delete this ?

File: packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts [hunk 3]
```diff
@@ -7,60 +7,34 @@
...
 
 vscode.workspace.onDidChangeTextDocument(e => {
   if (ForceSourceDeployExecutor.errorCollection.has(e.document.uri)) {
     ForceSourceDeployExecutor.errorCollection.delete(e.document.uri);
   }
 });
 
-export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<
-  SelectedPath
+export abstract class ForceSourceDeployExecutor extends SfdxCommandletExecutor<
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts]: Awesome, this makes every type of deploy command render any deploy errors in the problem view.

File: packages/salesforcedx-vscode-core/src/commands/forceSourceDeployManifest.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  SfdxWorkspaceChecker
+} from './commands';
+import { ForceSourceDeployExecutor } from './forceSourceDeploy';
+
+export class ForceSourceDeployManifestExecutor extends ForceSourceDeployExecutor {
+  public build(manifestPath: string): Command {
+    const commandBuilder = new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_source_deploy_text'))
+      .withArg('force:source:deploy')
+      .withLogName('force_source_deploy')
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceDeployManifest.ts]: Lets change this to `fource_source_deploy_manifest`

File: packages/salesforcedx-vscode-core/src/commands/forceSourceDeploySourcePath.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+} from './commands';
+import { ForceSourceDeployExecutor } from './forceSourceDeploy';
+import { SourcePathChecker } from './forceSourceRetrieveSourcePath';
+
+export class ForceSourceDeploySourcePathExecutor extends ForceSourceDeployExecutor {
+  public build(sourcePath: string): Command {
+    const commandBuilder = new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_source_deploy_text'))
+      .withArg('force:source:deploy')
+      .withLogName('force_source_deploy')
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceDeploySourcePath.ts]: Lets change this to `force_source_deploy_with_sourcepath` to keep consistency with the log ids for the retrieve commands.

File: packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        for (const packagePath of fullPackagePaths) {
+          if (sourcePath.startsWith(packagePath)) {
+            sourcePathIsInPackageDirectory = true;
+            break;
+          }
+        }
+        if (sourcePathIsInPackageDirectory) {
+          return inputs;
+        }
+      } catch (error) {}
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts]: We should log this error and/or send it to our telemetry service.

File: packages/system-tests/assets/sfdx-simple/.vscode/settings.json (hunk 2 above)
Comment (allileong) [reply packages/system-tests/assets/sfdx-simple/.vscode/settings.json]: Thanks for catching this. I did not mean to delete this and will put it back.

File: packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+          sfdxProjectPath
+        );
+        const fullPackagePaths = packageDirectoryPaths.map(
+          packageDirectoryPath =>
+            path.join(sfdxProjectPath, packageDirectoryPath)
+        );
+
+        let sourcePathIsInPackageDirectory = false;
+        for (const packagePath of fullPackagePaths) {
+          if (sourcePath.startsWith(packagePath)) {
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts]: Wouldn't `packagePath` be something like `force-app` and `sourcePath` be something like `Users/myuser/repo_name/force-app` ?

Comment (lcampos) [reply packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts]: Disregard this, I've checked and I was wrong 👍

**Pull request merged**